romancal/datamodels/library.py

```diff
@@ -0,0 +1,532 @@
+
+        # make sure first model is loaded in memory (as expected by stpipe)
+        if self._asn_n_members == 1:
+            # FIXME stpipe also reaches into _models (instead of _model_store)
+            self._models = [self._load_member(0)]
+
+    def __del__(self):
+        # FIXME when stpipe no longer uses '_models'
+        if hasattr(self, "_models"):
+            self._models[0].close()
+
+    @property
+    def asn(self):
+        # return a "read only" association
+        def _to_read_only(obj):
+            if isinstance(obj, dict):
+                return MappingProxyType(obj)
+            if isinstance(obj, list):
+                return tuple(obj)
+            return obj
+
+        return asdf.treeutil.walk_and_modify(self._asn, _to_read_only)
+
+    # TODO we may want to not expose this as it could go out-of-sync
+    # pretty easily with the actual models.
+    # @property
+    # def members(self):
+    #     return self.asn['products'][0]['members']
+
+    @property
+    def group_names(self):
+        names = set()
+        for member in self._members:
+            names.add(member["group_id"])
+        return names
+
+    @property
+    def group_indices(self):
+        group_dict = {}
+        for i, member in enumerate(self._members):
+            group_id = member["group_id"]
+            if group_id not in group_dict:
+                group_dict[group_id] = []
+            group_dict[group_id].append(i)
+        return group_dict
+
+    def __len__(self):
+        return len(self._members)
+
+    def __getitem__(self, index):
+        if not self._open:
+            raise ClosedLibraryError("ModelLibrary is not open")
+
+        # if model was already borrowed, raise
+        if index in self._ledger:
+            raise BorrowError("Attempt to double-borrow model")
+
+        if index in self._model_store:
+            model = self._model_store[index]
+        else:
+            model = self._load_member(index)
+            if not self._on_disk:
+                # it's ok to keep this in memory since _on_disk is False
+                self._model_store[index] = model
+
+        # track the model is "in use"
+        self._ledger[index] = model
+        return model
+
+    def __setitem__(self, index, model):
+        if index not in self._ledger:
+            raise BorrowError("Attempt to return non-borrowed model")
+
+        # un-track this model
+        del self._ledger[index]
+
+        # and store it
+        self._model_store[index] = model
+
+        # TODO should we allow this to change group_id for the member?
+
+    def discard(self, index, model):
+        # TODO it might be worth allowing `discard(model)` by adding
+        # an index of {id(model): index} to the ledger to look up the index
+        if index not in self._ledger:
+            raise BorrowError("Attempt to discard non-borrowed model")
+
+        # un-track this model
+        del self._ledger[index]
+        # but do not store it
```

Comment:
Maybe just repeating Ned's comment, discarding is a pretty common case in your code and currently you always do discard(i, model).  But I think at present discard(i, model) doesn't look at model?  In that case I would definitely make discard(...) take either i or model but not both.  The current syntax kind of makes you use:
```
model = library[i]
...
library.discard(model, i)
```
instead of
```
function(library[i])
library.discard(i)
```

It also made me ask myself "what is it doing with that model I just gave back to it that I thought I wanted to throw away?!"

Reply:
"throw away" might be another name to consider :)

I am starting to lean towards `shelve` (with a `modify` argument that defaults to `True`) this would replace `discard` with something like `library.shelve(model, modify=False)`. One potential downside to this is that internally the library operates on the index of the model within the container (so that references to the model don't need to be kept). Supporting either `discard(model)` or `shelve(model)` would require keeping some mapping of model to index. This could be done with weak references but that would require a more complex library. We could instead keep references the model and accept that between a `borrow` and a `shelve` even if the step code frees all references the model would stay in memory because of the library.